src/dettrace.cpp

```diff
@@ -0,0 +1,522 @@
+        personality(PER_LINUX | ADDR_NO_RANDOMIZE), "Unable to disable ASLR");
+  }
+
+  if (opts.mount.chroot_dir &&
+      (opts.clone_ns_flags & CLONE_NEWNS) == CLONE_NEWNS) {
+    const auto pathToChroot = std::string{opts.mount.chroot_dir};
+
+    if (!fileExists("/dev/null")) {
+      // we're running under reprotest as sudo, so we can use real mknod
+      // hat tip to:
+      // https://unix.stackexchange.com/questions/27279/how-to-create-dev-null
+      dev_t dev = makedev(1, 3);
+      mode_t mode =
+          S_IFCHR | S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH;
+      doWithCheck(mknod("/dev/null", mode, dev), "mknod");
+    }
+
+    if (opts.mount.with_devrand_overrides) {
+      createFileIfNotExist("/dev/random");
+      mountDir(devrandFifoPath, "/dev/random");
+      createFileIfNotExist("/dev/urandom");
+      mountDir(devUrandFifoPath, "/dev/urandom");
+    }
+
+    if (opts.mount.with_proc_overrides) {
+      mountDir(pathToChroot + "/proc/meminfo", "/proc/meminfo");
+      mountDir(pathToChroot + "/proc/stat", "/proc/stat");
+      mountDir(pathToChroot + "/proc/filesystems", "/proc/filesystems");
+    }
+
+    if (opts.mount.with_etc_overrides) {
+      mountDir(pathToChroot + "/etc/hosts", "/etc/hosts");
+      mountDir(pathToChroot + "/etc/passwd", "/etc/passwd");
+      mountDir(pathToChroot + "/etc/group", "/etc/group");
+      mountDir(pathToChroot + "/etc/ld.so.cache", "/etc/ld.so.cache");
+    }
+
+    // for (auto v : args->volume) {
+    //  mountDir(v.source, v.target);
+    //}
+
+    // this have to be done before mount /dev/{u}random because the source file
+    // is under previous /tmp
+    doWithCheck(
+        mount("none", "/tmp", "tmpfs", 0, NULL), "mount /tmp as tmpfs failed");
+  }
+
+  // set working dir
+  if (opts.workdir) {
+    doWithCheck(
+        chdir(opts.workdir), std::string{"unable to chdir to "} + opts.workdir);
+  }
+
+  // trap on rdtsc/rdtscp insns
+  doWithCheck(
+      prctl(PR_SET_TSC, PR_TSC_SIGSEGV, 0, 0, 0), "Pre-clone prctl error");
+  doWithCheck(
+      prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0),
+      "Pre-clone prctl error: setting no new privs");
+
+  // Perform execve based on user command.
+  ptracer::doPtrace(PTRACE_TRACEME, 0, NULL, NULL);
+
+  // Set up seccomp + bpf filters using libseccomp.
+  // Default action to take when no rule applies to system call. We send a
+  // PTRACE_SECCOMP event message to the tracer with a unique data: INT16_MAX
+  seccomp myFilter{opts.debug_level, opts.convert_uids};
+
+  // Stop ourselves until the tracer is ready. This ensures the tracer has time
+  // to get set up.
+  raise(SIGSTOP);
+
+  myFilter.loadFilterToKernel();
+
+  // execvpe() duplicates the actions of the shell in searching for an
+  // executable file if the specified filename does not contain a slash (/)
+  // character.
+  int val = opts.envs ? execvpe(opts.program, opts.argv, opts.envs)
+                      : execvp(opts.program, opts.argv);
```

Comment:
Because we just loaded seccomp, it is more ideal to do raw syscall (`execve`) instead of library calls, because library calls may do other syscalls which might trigger ptrace SECCOMP event, which we're not ready to handle.